### PR TITLE
MONGOID-5617 - Increase performance of `Mongoid::Matcher#extract_attribute` by avoiding BSON::Document instantiation

### PR DESCRIPTION
**Pull request by jasonpenny**
https://jira.mongodb.org/browse/MONGOID-5617

`Mongoid::Matcher#extract_attribute` creates a `BSON::Document` for indifferent access of hash keys.

Performance can be improved by explicitly attempting string and symbol access.

```console
$ ruby -v
ruby 2.7.3p183 (2021-04-05 revision 6847ee089d) [arm64-darwin21]

$ bundle exec ruby perf/benchmark_matches.rb 2>/dev/null
Warming up --------------------------------------
_matches? with BSON::Document
                         9.424k i/100ms
_matches? with indifferent key methods
                         9.999k i/100ms
Calculating -------------------------------------
_matches? with BSON::Document
                         94.447k (± 0.8%) i/s -    480.624k in   5.089113s
_matches? with indifferent key methods
                         99.887k (± 0.7%) i/s -    499.950k in   5.005367s

Comparison:
_matches? with indifferent key methods:    99887.4 i/s
_matches? with BSON::Document:    94447.0 i/s - 1.06x  slower
```

### CODE REVIEW

File: perf/benchmark_matches.rb [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+SELECTORS = [
+  DocWithHashes.where(:"h.1.a".exists => false),
+  DocWithHashes.where(:"h.3.d".lt => "2021-04-19T23:05:34.142Z"),
+  DocWithHashes.where(:"h.4.d".exists => false).or(:"h.4.ct" => true),
+].map { |criteria| criteria.selector }
+
+Benchmark.ips do |bm|
+  bm.report("_matches? with BSON::Document") do
+    Mongoid::Matcher.instance_variable_set(:@attributes_as_bson_doc, true)
```
Comment (comandeo) [inline perf/benchmark_matches.rb:31]: If we remove the old implementation of the matcher, we should remove the comparison, too.

**Review by comandeo — CHANGES_REQUESTED** [review]
The PR looks very good, just some concerns should be addressed. Please let us know if you have capacity to clean the PR up.

File: lib/mongoid/matcher.rb [hunk 2]
```diff
@@ -3,6 +3,7 @@ module Mongoid
 
   # @api private
   module Matcher
+    @attributes_as_bson_doc = false
```
[the remark this answers is not in the document]
Comment (jasonpenny) [reply lib/mongoid/matcher.rb:6]: This was not intended to be kept, it was just for the benchmark

**Review by jamis — APPROVED** [review]
Looks great!

**Review by comandeo — APPROVED** [review]

**Comment by comandeo** [comment]
Thank you very much for your contribution @jasonpenny

**Pull request merged**